artemis-cli/src/main/java/org/apache/activemq/artemis/cli/commands/messages/perf/AsyncJms2ProducerFacade.java

```diff
@@ -0,0 +1,256 @@
+      } catch (final JMSException ex) {
+         throw new IllegalStateException(ex);
+      }
+      this.transactionCapacity = transactionCapacity;
+      this.pendingMsgInTransaction = 0;
+      this.completedMsgInTransaction = 0;
+      this.closing = false;
+      this.closed = false;
+   }
+
+   public long getId() {
+      return id;
+   }
+
+   public Destination getDestination() {
+      return destination;
+   }
+
+   BytesMessage createBytesMessage() throws JMSException {
+      return session.createBytesMessage();
+   }
+
+   private void addedPendingSend() {
+      if (transactionCapacity > 0 && pendingMsgInTransaction == transactionCapacity) {
+         throw new IllegalStateException("cannot add more pending send");
+      }
+      if (maxPending > 0 && pending == maxPending) {
+         throw new IllegalStateException("cannot add more pending");
+      }
+      pending++;
+      pendingMsgInTransaction++;
+   }
+
+   private boolean isAvailable() {
+      if (maxPending > 0 && pending == maxPending) {
+         return false;
+      }
+      return transactionCapacity == 0 || pendingMsgInTransaction != transactionCapacity;
+   }
+
+   public enum SendAttemptResult {
+      Closing, Closed, NotAvailable, Success
+   }
+
+   public SendAttemptResult trySend(final Message message,
+                                    final CompletionListener completionListener,
+                                    final Runnable availableObserver) throws JMSException {
+      if (closing) {
+         return SendAttemptResult.Closing;
+      }
+      if (closed) {
+         return SendAttemptResult.Closed;
+      }
+      if (!isAvailable()) {
+         availableObservers.add(availableObserver);
+         orderedIncrementNotAvailable();
+         return SendAttemptResult.NotAvailable;
+      }
+      producer.send(message, completionListener);
+      orderedIncrementSent();
+      addedPendingSend();
+      return SendAttemptResult.Success;
+   }
+
+   public void onSendErrored() {
+      if (closed) {
+         return;
+      }
+      availableObservers.clear();
+      closedObservers.forEach(Runnable::run);
+      closedObservers.clear();
+      closed = true;
+   }
+
+   public void onSendCompleted() {
+      if (closed) {
+         return;
+      }
+      orderedIncrementCompleted();
+      if (transactionCapacity > 0 && completedMsgInTransaction == transactionCapacity) {
+         throw new IllegalStateException("cannot complete more send");
+      }
+      if (pending == 0) {
+         throw new IllegalStateException("cannot complete more send");
+      }
+      pending--;
+      completedMsgInTransaction++;
+      if (transactionCapacity > 0) {
+         if (completedMsgInTransaction == transactionCapacity || (closing && pending == 0)) {
+            completedMsgInTransaction = 0;
+            pendingMsgInTransaction = 0;
+            try {
+               session.commit();
+            } catch (final Throwable t) {
+               // TODO log
```

Comment:
Seems fairly important to do this at least. Not immediately clear why it wouldnt just throw. What if it was a TransactionRolledBackException?